src/camera.py

```diff
@@ -0,0 +1,152 @@
+import logging
+import os
+from typing import Dict, Optional
+
+import cv2
+import requests  # type: ignore
+from lxml.html import fromstring
+from numpy import std
+from requests.exceptions import ReadTimeout, RequestException  # type: ignore
+
+import settings as s
+from exceptions import FetchCamerasException, SaveImageException
+from shared import exponential_backoff
+
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger(__name__)
+
+
+class Camera:
+    """Camera class to handle camera-related operations."""
+
+    def __init__(self, url: str):
+        self.url = url
+        self.id = self._get_camera_id()
+        self.page_content = self._get_camera_page(s.REQUEST_HEADERS)
+        self.page_tree = fromstring(self.page_content) if self.page_content else None
+        self.stream_url = self._find_camera_url() if self.page_tree is not None else None
+        self.details = self._get_camera_details() if self.page_tree is not None else None
+        self.info = self._parse_camera_details() if self.details is not None else None
+
+    def _get_camera_id(self) -> str:
+        """Extracts the camera ID from the URL."""
+        return "".join(char for char in self.url if char.isdigit())
+
+    def _get_camera_page(self, request_headers: Dict[str, str]) -> Optional[bytes]:
+        """Fetches the camera page content."""
+        try:
+            r = requests.get(self.url, headers=request_headers)
+            return r.content
+        except (RequestException, OSError) as e:
+            logger.error("error capturing image: " + str(e))
+            return None
+
+    def _find_camera_url(self) -> Optional[str]:
+        """Finds the camera stream URL."""
+        if self.page_tree is not None:
+            elems = self.page_tree.xpath("//img/@src")
+        if len(elems) > 0:
+            return elems[0]
+        return None
+
+    def _url_is_valid(self) -> bool:
+        """Checks if the stream URL is valid."""
+        if self.stream_url is None:
+            return False
+
+        return all(
+            [
+                self.stream_url != "/static/no.jpg",
+                "?stream" not in self.stream_url,
+                ".jpg" in self.stream_url,
+            ]
+        )
+
+    def _get_camera_details(self) -> Optional[str]:
+        """Extracts the camera details."""
+        details = self.page_tree.xpath('//div[@class="camera-details"]')
+        details_tuple = tuple(detail.text_content() for detail in details)
+        details = "".join(detail.replace("\n", "").replace("\t", "").strip() for detail in details_tuple)
+        return details
+
+    def _parse_camera_details(self) -> Optional[Dict[str, str]]:
+        """Parses the camera details and returns the camera info as a dictionary."""
+
+        if not self.details:
+            return None
+
+        details = self.details
+

```

Comment:
It's not completely clear to me what all these find() operations are doing here. What is the problem this block solves? I may be able to recommend a less repetitive method.

Reply:
The original code used the `find()` function repeatedly to figure out where each piece of info was in the string, then sliced the string to get what it needed.

I rewrote the function to now loop over a list of the info I want and use `find()` inside the loop to get the start and end positions for each piece. This way, I don't need to write the same code over and over again for each piece of info.

```python
    def _parse_camera_details(self) -> Dict[str, str]:
        """Parses the camera details and returns the camera info as a dictionary."""

        details = self.details if self.details is not None else None
        camera_info = {}

        if details is not None:
            for key in ["City", "Region", "Country", "Country code"]:
                start = details.find(f"{key}: ") + len(f"{key}: ")
                end = details.find(" ", start)
                value = details[start:end].strip()
                camera_info[key] = value

        return camera_info
```

Reply:
I've modified this further to correctly parse the details string by searching for the correct delimiters (newlines) and not relying on a specific order of keys. This ensures that the function the details string is parsed accurately regardless of the order of keys or the presence of spaces within values.

```python
    def _parse_camera_details(self) -> Optional[Dict[str, str]]:
        """Parses the camera details and returns the camera info as a dictionary."""
        details = self.details if self.details is not None else None
        camera_info = {}

        if details is not None:
            keys = [("City", "Latitude"), ("Region", "City"), ("Country", "Country code"), ("Country code", "Region")]
            for key, next_key in keys:
                start = details.find(f"{key}:") + len(f"{key}:")
                end = details.find(f"{next_key}:")
                value = details[start:end].strip()
                camera_info[key] = value
        return camera_info
```